### PR TITLE
Update codebase to new cuDF API

### PR DESCRIPTION
**Pull request by gaohao95**
This PR intends to implement #6.

### CODE REVIEW

**Comment by nsakharnykh** [comment]
@jrhemstad can you take a look at this PR when you have time? mostly cuDF changes to update to the new libcudf++ API

File: src/comm.cuh [hunk 1]
```diff
@@ -158,25 +165,24 @@ std::vector<comm_handle_t> recv_data_by_offset(
...
-void* merge_free_received_offset(
-                                 std::vector<void *> received_data,
-                                 const std::vector<int64_t> &bucket_count,
-                                 size_t item_size,
-                                 int64_t &total_count,
-                                 Communicator *communicator=nullptr,
-                                 bool self_free=true)
+rmm::device_buffer
+merge_free_received_offset(
+    std::vector<void *> received_data,
```
Comment (jrhemstad) [inline src/comm.cuh]: Passing a vector by value looks like a mistake to me.

File: src/generate_table.cuh [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+#include "error.cuh"
+#include "../generate_dataset/generate_dataset.cuh"
+#include "distributed_join.cuh"
+
+using cudf::experimental::table;
+
+
+template <typename col_type>
+cudf::data_type gdf_dtype_from_col_type()
```
Comment (jrhemstad) [inline src/generate_table.cuh]: Use `cudf::type_to_id``.

File: src/generate_table.cuh [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+    return std::make_pair(std::move(build_table), std::move(probe_table));
+}
+
+
+template <typename data_type>
+void add_constant_to_column(cudf::mutable_column_view column, data_type constant)
+{
+    auto buffer_ptr = thrust::device_pointer_cast(column.head<data_type>());
+
+    thrust::for_each(buffer_ptr, buffer_ptr + column.size(),
```
Comment (jrhemstad) [inline src/generate_table.cuh]: `thrust::transform`.

File: src/generate_table.cuh [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+
+    // Generate build and probe table data
+
+    generate_input_tables<KEY_T, cudf::size_type>(
+        build[0]->mutable_view().head<KEY_T>(), build_table_nrows,
+        probe[0]->mutable_view().head<KEY_T>(), probe_table_nrows,
+        selectivity, rand_max, uniq_build_tbl_keys
+    );
+
+    linear_sequence<PAYLOAD_T, cudf::size_type><<<(build_table_nrows+127)/128,128>>>(
```
Comment (jrhemstad) [inline src/generate_table.cuh]: Use `thrust::sequence`: https://thrust.github.io/doc/group__transformations_ga9d17b70466df6b9a57f94170935a03f8.html

File: src/distributed_join.cuh [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+        );
+    }
+
+    return std::make_unique<table>(std::move(recv_columns));
+}
+
+
+void
+inner_join_func(
+    vector<vector<vector<void *> > > &left_buckets,
```
Comment (jrhemstad) [inline src/distributed_join.cuh]: All of these non-const vectors looks fishy to me.

File: benchmark/distributed_join.cu [hunk 6]
```diff
@@ -14,22 +14,26 @@
...
+
+#include "../src/communicator.h"
 #include "../src/error.cuh"
-#include "../src/comm.cuh"
+#include "../src/generate_table.cuh"
+#include "../src/distributed_join.cuh"
 
-#define BUILD_TABLE_SIZE_EACH_RANK 100'000'000
-#define PROBE_TABLE_SIZE_EACH_RANK 100'000'000
+#define BUILD_TABLE_NROWS_EACH_RANK 100'000'000
```
[the remark this answers is not in the document]
Comment (gaohao95) [reply benchmark/distributed_join.cu]: Good point. Fixed in https://github.com/rapidsai/distributed-join/pull/10/commits/d13f6a8433c42adbb6321f543f3ed0c9842e8e9d.

File: src/comm.cuh (hunk 1 above)
Comment (gaohao95) [reply src/comm.cuh]: Fixed in https://github.com/rapidsai/distributed-join/pull/10/commits/dd3fe11b04d1bf08930ab7507ab8a10a6ce6f60c

File: src/generate_table.cuh (hunk 2 above)
Comment (gaohao95) [reply src/generate_table.cuh]: Done.

File: src/distributed_join.cuh (hunk 5 above)
Comment (gaohao95) [reply src/distributed_join.cuh]: Added `const` qualifiers to them in https://github.com/rapidsai/distributed-join/pull/10/commits/9ce2080062df4b8d58f2524be4c601f321328e1a

**Comment by gaohao95** [comment]
@jrhemstad Thanks a lot for the review. I really appreciate these feedback.

**Comment by gaohao95** [comment]
> Should this be a `const&`?

`const` is added in a following commit https://github.com/rapidsai/distributed-join/pull/10/commits/9ce2080062df4b8d58f2524be4c601f321328e1a#diff-492eca7c4eff57c31bfa5be3fb67f18fR170

File: src/comm.cuh [hunk 7]
```diff
@@ -72,12 +75,13 @@ MPI_Datatype mpi_dtype_from_c_type()
...
-std::vector<comm_handle_t> send_data_by_offset(
-                                               const void *data,
-                                               const int *offset,
-                                               size_t item_size,
-                                               Communicator *communicator,
-                                               bool self_send=true)
+std::vector<comm_handle_t>
+send_data_by_offset(
+    const void *data,
+    std::vector<int> offset,
```
Comment (jrhemstad) [inline src/comm.cuh]:
```suggestion
    std::vector<int> const& offset,
```

File: src/comm.cuh (hunk 1 above)
Comment (jrhemstad) [inline src/comm.cuh]: The latest state of the code, this is still being passed by value. 
```suggestion
    std::vector<void *> const& received_data,
```

Comment (gaohao95) [reply src/comm.cuh]: Sorry if I missed anything, but I am really confused. I think I am passing it by reference at the latest state of the code [here](https://github.com/rapidsai/distributed-join/blob/9ce2080062df4b8d58f2524be4c601f321328e1a/src/comm.cuh#L170)?

Comment (jrhemstad) [reply src/comm.cuh]: I don't know how, but somehow I was reviewing a stale version of the code.

**Comment by jrhemstad** [comment]
Do you have updated performance numbers after these changes?

**Comment by gaohao95** [comment]
> Do you have updated performance numbers after these changes?

Not yet. We will test it after we figure out some communication layer configurations (https://github.com/rapidsai/distributed-join/pull/11).

**Comment by gaohao95** [comment]
https://github.com/rapidsai/cudf/pull/3748 is recently merged after 4d0976d, which is based on [our initial PR](https://github.com/rapidsai/cudf/pull/2623). This may improve hash partition performance and align with my custom branch.

**Comment by nsakharnykh** [comment]
I tested with today's master `ac143f10236b32842bfac7709a74477c2f8d0df6` but got similar timings, a little faster than before rapidsai/cudf#3748 got merged, but still slower than `branch-0.8`: 
```$ mpirun -n 8 --bind-to none --mca pml ucx --mca btl ^openib,smcuda benchmark/run_sample.sh
rank 0 gpu list 0,1,2,3,4,5,6,7 cpu bind 1-4 ndev mlx5_0:1
rank 1 gpu list 0,1,2,3,4,5,6,7 cpu bind 5-8 ndev mlx5_0:1
rank 2 gpu list 0,1,2,3,4,5,6,7 cpu bind 10-13 ndev mlx5_1:1
rank 3 gpu list 0,1,2,3,4,5,6,7 cpu bind 15-18 ndev mlx5_1:1
rank 4 gpu list 0,1,2,3,4,5,6,7 cpu bind 21-24 ndev mlx5_2:1
rank 5 gpu list 0,1,2,3,4,5,6,7 cpu bind 25-28 ndev mlx5_2:1
rank 6 gpu list 0,1,2,3,4,5,6,7 cpu bind 30-33 ndev mlx5_3:1
rank 7 gpu list 0,1,2,3,4,5,6,7 cpu bind 35-38 ndev mlx5_3:1
Device count: 8
Rank 0 select 0/8 GPU
Device count: 8
Rank 6 select 6/8 GPU
Device count: 8
Rank 1 select 1/8 GPU
Device count: 8
Rank 5 select 5/8 GPU
Device count: 8
Rank 4 select 4/8 GPU
Device count: 8
Rank 7 select 7/8 GPU
Device count: 8
Rank 2 select 2/8 GPU
Device count: 8
Rank 3 select 3/8 GPU
Elasped time (s) 0.607329
```

**Comment by jrhemstad** [comment]
It's not too surprising that things got a bit slower. There was some changes necessary in the join algorithm to add string support which may have slowed things down, but we'd need to see profiles to know exactly what the root cause is.

**Comment by gaohao95** [comment]
Merging since we can get as good or even better performance than last year's number on legacy cuDF.

**Pull request merged**